vault/plugin_catalog.go

```diff
@@ -64,18 +64,11 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
 // type. It will first attempt to run as a database plugin then a backend
 // plugin. Both of these will be run in metadata mode.
 func (c *PluginCatalog) getPluginTypeFromUnknown(ctx context.Context, logger log.Logger, plugin *pluginutil.PluginRunner) (consts.PluginType, error) {
-
-	{
-		// Attempt to run as database plugin
-		client, err := dbplugin.NewPluginClient(ctx, nil, plugin, log.NewNullLogger(), true)
-		if err == nil {
-			// Close the client and cleanup the plugin process
-			client.Close()
-			return consts.PluginTypeDatabase, nil
-		} else {
-			logger.Warn(fmt.Sprintf("received %s attempting as db plugin, attempting as auth/secret plugin", err))
-		}
+	err := isDatabasePlugin(ctx, plugin, logger)
+	if err == nil {
+		return consts.PluginTypeDatabase, nil
 	}
+	logger.Warn(fmt.Sprintf("received %s attempting as db plugin, attempting as auth/secret plugin", err))
```

Comment:
The error might look a bit off in here now since it can be a multierror, so consider appending error to the end of the string. https://play.golang.org/p/rOofMYebB2d

Reply:
I ended up keeping this warning since it was there in the previous version, however I don't think this message makes sense as a warning. Every plugin goes through this path, and if the plugin isn't a database plugin you're going to get a warning saying the plugin isn't a database plugin. What do you think of changing this to a debug (or trace?) log? I'm also tempted to not include the error message at all as I'm skeptical that it's useful in this context. Thoughts?

Reply:
`getPluginTypeFromUnknown` shouldn't be called that often -- once when setting up the plugin catalog if a plugin has not been upgraded yet, and when a plugin with unknown type is added to the catalog (which shouldn't happen since we're now [enforcing the type](https://www.vaultproject.io/api-docs/system/plugins-catalog#register-plugin) to be provided via the API). The latter is more of a backward compat behavior, so new plugin registration operation shouldn't be hitting this path.

Reply:
For posterity:
@calvn & I talked offline and agreed that we should keep this as a warning, but move it to the end of the function if we are unable to determine the type of the plugin. I also changed it so the message is a bit more descriptive differentiating between an unknown type vs one that isn't recognized by this particular function. That will help guard against adding new types to the `logical.BackendType` enum.